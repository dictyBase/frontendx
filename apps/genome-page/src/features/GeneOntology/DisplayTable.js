// @flow
import React, { Component } from "react"
import { Table, Column } from "react-virtualized"
import "react-virtualized/styles.css"

const BASE_ROW_HEIGHT = 30
const MAX_NAME_CHARS_PER_LINE = 20

type Props = {
    /** Width of table cell */
    cellWidth: number,
    /** Height of table cell */
    cellHeight: number,
    /** Height of table */
    height: number,
    /** The data fetched from the API */
    data: Object
}

/**
 * React Virtualized table that displays relevant GO data
 */

export default class DisplayTable extends Component<Props> {
    static defaultProps = {
        cellWidth: 130,
        cellHeight: 60,
        height: 330
    }

    getRowHeight = ({ index }: { index: number }) => {
        const data = this.props.data[index]
        const numLines = Math.ceil(data.term.length / MAX_NAME_CHARS_PER_LINE)

        // if (numLines > MAX_NAME_CHARS_PER_LINE) {
        //     return numLines * BASE_ROW_HEIGHT
        // } else {
        //     return BASE_ROW_HEIGHT
        // }

        return numLines * BASE_ROW_HEIGHT
    }
    getRowStyle = ({ index }: { index: number }) => {
        if (index === -1) {
            return {
                margin: "0 auto",
                borderTop: "1px solid #efefef",
                borderBottom: "1px solid #efefef"
            }
        } else if (index === this.props.data.length) {
            return {}
        } else if (index % 2 > 0) {
            return {
                borderBottom: "1px solid #efefef",
                backgroundColor: "#eaf2ff"
            }
        } else if (index % 2 === 0) {
            return {
                borderBottom: "1px solid #efefef"
            }
        }
    }
    rowGetter = ({ index }: { index: number }) => {
        const data = this.props.data
        if (data[index]) {
            return data[index]
        }
    }
    cellDataGetter = ({
        rowData,
        dataKey
    }: {
        rowData: Object,
        dataKey: string
    }) => {
        if (rowData) {
            return rowData[dataKey]
        }
    }
    descriptorRenderer = ({
        rowData,
        cellData
    }: {
        rowData: Object,
        cellData: any
    }) => {
        if (rowData) {
            return <div style={{ whiteSpace: "normal" }}>{cellData}</div>
        }
    }

    render() {
        const rowCount = this.props.data.length
        const { cellWidth, cellHeight } = this.props

        return (
            <div className="wrapper">
                <Table
                    headerHeight={50}
                    height={cellHeight * 7}
                    width={cellWidth * 3 + 250 + 250}
                    rowCount={rowCount}
                    rowGetter={this.rowGetter}
                    rowHeight={this.getRowHeight}
                    rowStyle={this.getRowStyle}>
                    <Column
                        label="GO term + Extension"
                        dataKey="term"
                        width={250}
                        cellDataGetter={this.cellDataGetter}
                        cellRenderer={this.descriptorRenderer}
                    />
                    <Column label="Evidence" dataKey="evidence" width={100} />
                    <Column
                        label="With"
                        dataKey="with"
                        width={150}
                        cellDataGetter={this.cellDataGetter}
                        cellRenderer={this.descriptorRenderer}
                    />
                    <Column
                        label="Reference"
                        dataKey="reference"
                        width={250}
                        cellDataGetter={this.cellDataGetter}
                        cellRenderer={this.descriptorRenderer}
                    />
                    <Column label="Date" dataKey="date" width={100} />
                    <Column label="Source" dataKey="source" width={100} />
                </Table>
            </div>
        )
    }
}
