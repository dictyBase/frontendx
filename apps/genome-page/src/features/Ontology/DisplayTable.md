```js
const data = require("./mockData").data
;<DisplayTable goaData={data} />
```
