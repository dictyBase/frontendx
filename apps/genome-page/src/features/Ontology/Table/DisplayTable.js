// @flow
import React, { Component, Fragment } from "react"
import { withStyles } from "@material-ui/core/styles"
import Table from "@material-ui/core/Table"
import TableBody from "@material-ui/core/TableBody"
import TableCell from "@material-ui/core/TableCell"
import TableRow from "@material-ui/core/TableRow"
import Paper from "@material-ui/core/Paper"

import EnhancedTableHead from "./EnhancedTableHead"
import pubLinkGenerator from "../utils/pubLinkGenerator"
import evidenceLinkGenerator from "../utils/evidenceLinkGenerator"
import withLinkGenerator from "../utils/withLinkGenerator"
import removeUnderscores from "../utils/removeUnderscores"
import dateConverter from "../utils/dateConverter"
import qualifierFormatter from "../utils/qualifierFormatter"

const styles = (theme: Object) => ({
  root: {
    width: "100%",
    overflowX: "auto",
  },
  row: {
    "&:nth-of-type(even)": {
      backgroundColor: theme.palette.background.default,
    },
  },
  link: {
    textDecoration: "none",
    color: "#4C5E81",
    "&:visited": {
      color: "#4C5E81",
    },
    "&:hover": {
      textDecoration: "underline",
    },
  },
})

// helper function for table sorting
const getSorting = (order, orderBy) =>
  order === "desc"
    ? (a, b) => (b[orderBy] < a[orderBy] ? -1 : 1)
    : (a, b) => (a[orderBy] < b[orderBy] ? -1 : 1)

type Props = {
  /** Material-UI styling */
  classes: Object,
  /** GOA data fetched from QuickGO API */
  goaData: Array<Object>,
}

type State = {
  /** The order to sort the column */
  order: string,
  /** The item to be ordered by */
  orderBy: string,
}

/**
 * The display table used inside each panel in the GO tabs.
 */

class DisplayTable extends Component<Props, State> {
  state = {
    order: "asc",
    orderBy: "evidence_code",
  }

  handleRequestSort = (event, property) => {
    const orderBy = property
    let order = "desc"

    if (this.state.orderBy === property && this.state.order === "desc") {
      order = "asc"
    }

    this.setState({ order, orderBy })
  }

  render() {
    const { order, orderBy } = this.state
    const { classes, goaData } = this.props

    // get array for item.attributes, helps for sorting table
    const goaDataAttributes = goaData.map((item: Object) => item.attributes)

    return (
      <Paper className={classes.root}>
        <Table>
          <colgroup>
            <col style={{ width: "25%" }} />
            <col style={{ width: "10%" }} />
            <col style={{ width: "20%" }} />
            <col style={{ width: "25%" }} />
            <col style={{ width: "15%" }} />
            <col style={{ width: "10%" }} />
          </colgroup>
          <EnhancedTableHead
            order={order}
            orderBy={orderBy}
            onRequestSort={this.handleRequestSort}
          />
          <TableBody>
            {goaDataAttributes
              .sort(getSorting(order, orderBy))
              .map((item: Object, index: number) => (
                <TableRow className={classes.row} key={index}>
                  <TableCell component="th" scope="row">
                    {qualifierFormatter(item.qualifier)} {item.goterm}
                  </TableCell>
                  <TableCell>
                    <a
                      className={classes.link}
                      href={evidenceLinkGenerator(item.evidence_code)}
                      target="_blank">
                      {item.evidence_code}
                    </a>
                  </TableCell>
                  <TableCell>
                    {item.with !== null &&
                      item.with.map((item: Object) =>
                        item.connectedXrefs.map((xref: Object, i: string) => (
                          <Fragment key={i}>
                            <span>
                              <a
                                className={classes.link}
                                href={withLinkGenerator(xref.id)}
                                target="_blank">
                                {xref.db}:{xref.id}
                              </a>
                            </span>
                            <br />
                          </Fragment>
                        )),
                      )}
                  </TableCell>
                  <TableCell>
                    {item.extensions !== null &&
                      item.extensions.map((item: Object) =>
                        item.connectedXrefs.map((xref: Object, i: string) => (
                          <Fragment key={i}>
                            <span>
                              {" "}
                              <em>{removeUnderscores(xref.relation)}</em>{" "}
                              <a
                                className={classes.link}
                                href={withLinkGenerator(xref.id)}
                                target="_blank">
                                ({xref.db}:{xref.id})
                              </a>
                            </span>
                            <br />
                          </Fragment>
                        )),
                      )}
                  </TableCell>
                  <TableCell>
                    <a
                      className={classes.link}
                      href={pubLinkGenerator(item.publication)}
                      target="_blank">
                      {item.publication}
                    </a>
                  </TableCell>
                  <TableCell>{dateConverter(item.date)}</TableCell>
                </TableRow>
              ))}
          </TableBody>
        </Table>
      </Paper>
    )
  }
}

export default withStyles(styles)(DisplayTable)
