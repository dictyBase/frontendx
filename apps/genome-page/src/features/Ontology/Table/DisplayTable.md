```jsx
const data = require("../mockData").data

<DisplayTable data={data} />
```
