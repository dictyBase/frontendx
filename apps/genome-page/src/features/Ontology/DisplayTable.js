// @flow
import React from "react"
import { withStyles } from "@material-ui/core/styles"
import Table from "@material-ui/core/Table"
import TableBody from "@material-ui/core/TableBody"
import TableCell from "@material-ui/core/TableCell"
import TableHead from "@material-ui/core/TableHead"
import TableRow from "@material-ui/core/TableRow"
import Paper from "@material-ui/core/Paper"

const styles = (theme: Object) => ({
  root: {
    width: "100%",
    overflowX: "auto",
  },
  head: {
    backgroundColor: "#DFE8F6",
  },
  headerCell: {
    color: "#333",
    fontWeight: "600",
  },
  row: {
    "&:nth-of-type(even)": {
      backgroundColor: theme.palette.background.default,
    },
  },
})

type Props = {
  /** Material-UI styling */
  classes: Object,
  /** GOA data fetched from QuickGO API */
  goaData: Array<Object>,
}

/**
 * The display table used inside each panel in the GO tabs.
 */

const DisplayTable = (props: Props) => {
  const { classes, goaData } = props

  return (
    <Paper className={classes.root}>
      <Table>
        <colgroup>
          <col style={{ width: "25%" }} />
          <col style={{ width: "10%" }} />
          <col style={{ width: "25%" }} />
          <col style={{ width: "20%" }} />
          <col style={{ width: "10%" }} />
          <col style={{ width: "15%" }} />
        </colgroup>
        <TableHead className={classes.head}>
          <TableRow>
            <TableCell className={classes.headerCell}>
              GO Term + Extension
            </TableCell>
            <TableCell className={classes.headerCell}>Evidence</TableCell>
            <TableCell className={classes.headerCell}>With</TableCell>
            <TableCell className={classes.headerCell}>Reference</TableCell>
            <TableCell className={classes.headerCell}>Date</TableCell>
            <TableCell className={classes.headerCell}>Source</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {goaData.map((item: Object, index: string) => {
            return (
              <TableRow className={classes.row} key={index}>
                <TableCell component="th" scope="row">
                  {item.attributes.goterm}
                </TableCell>
                <TableCell>{item.attributes.evidence_code}</TableCell>
                <TableCell>
                  {item.attributes.with !== null &&
                    item.attributes.with.map((item: Object) => {
                      return item.connectedXrefs.map(
                        (xref: Object, i: string) => {
                          if (item.connectedXrefs.length === i + 1) {
                            return `${xref.db}:${xref.id}`
                          }
                          return `${xref.db}:${xref.id}, `
                        },
                      )
                    })}
                </TableCell>
                <TableCell>{item.attributes.publication}</TableCell>
                <TableCell>{item.attributes.date}</TableCell>
                <TableCell>{item.attributes.assigned_by}</TableCell>
              </TableRow>
            )
          })}
        </TableBody>
      </Table>
    </Paper>
  )
}

export default withStyles(styles)(DisplayTable)
