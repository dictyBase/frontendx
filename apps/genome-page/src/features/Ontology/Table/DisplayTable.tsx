import React from "react"
import { makeStyles } from "@material-ui/core/styles"
import Table from "@material-ui/core/Table"
import TableBody from "@material-ui/core/TableBody"
import TableCell from "@material-ui/core/TableCell"
import TableRow from "@material-ui/core/TableRow"
import Paper from "@material-ui/core/Paper"
// import EnhancedTableHead from "./EnhancedTableHead"
import ExtensionsDisplay from "./ExtensionsDisplay"
import WithDisplay from "./WithDisplay"
import pubLinkGenerator from "../utils/pubLinkGenerator"
import evidenceLinkGenerator from "../utils/evidenceLinkGenerator"
import dateConverter from "../utils/dateConverter"
import qualifierFormatter from "../utils/qualifierFormatter"
import sourceLinkGenerator from "../utils/sourceLinkGenerator"
// import getSorting from "./utils/getSorting"
import { GeneGOA } from "common/@types/gene-data"

const useStyles = makeStyles({
  root: {
    width: "100%",
    overflowX: "auto",
  },
  row: {
    "&:nth-of-type(even)": {
      // backgroundColor: theme.palette.background.default,
    },
  },
  link: {
    textDecoration: "none",
    color: "#4C5E81",
    "&:visited": {
      color: "#4C5E81",
    },
    "&:hover": {
      textDecoration: "underline",
    },
  },
  head: {
    backgroundColor: "#e6f2ff",
  },
  headerCell: {
    color: "#333",
    fontWeight: 600,
  },
})

type Props = {
  /** GO Annotations */
  data: Array<GeneGOA>
}

/**
 * The display table used inside each panel in the GO tabs.
 */

const DisplayTable = ({ data }: Props) => {
  const classes = useStyles()

  return (
    <Paper className={classes.root}>
      <Table>
        <colgroup>
          <col style={{ width: "25%" }} />
          <col style={{ width: "20%" }} />
          <col style={{ width: "10%" }} />
          <col style={{ width: "15%" }} />
          <col style={{ width: "10%" }} />
          <col style={{ width: "10%" }} />
          <col style={{ width: "10%" }} />
        </colgroup>
        {/* <EnhancedTableHead
            order={goa.tableOrder}
            orderBy={goa.tableSortBy}
            onRequestSort={this.handleRequestSort}
          /> */}
        <TableBody>
          {data
            // .sort(getSorting(goa.tableOrder, goa.tableSortBy))
            .map((item: GeneGOA, index: number) => (
              <TableRow className={classes.row} key={index}>
                <TableCell component="th" scope="row">
                  {qualifierFormatter(item.qualifier)} {item.go_term}
                </TableCell>

                <TableCell>
                  <ExtensionsDisplay extensions={item.extensions} />
                </TableCell>

                <TableCell>
                  <a
                    className={classes.link}
                    href={evidenceLinkGenerator(item.evidence_code)}
                    target="_blank"
                    rel="noopener noreferrer">
                    {item.evidence_code}
                  </a>
                </TableCell>

                <TableCell>
                  <WithDisplay withData={item.with} />
                </TableCell>

                <TableCell>
                  <a
                    className={classes.link}
                    href={pubLinkGenerator(item.publication)}
                    target="_blank"
                    rel="noopener noreferrer">
                    {item.publication}
                  </a>
                </TableCell>

                <TableCell>{dateConverter(item.date)}</TableCell>

                <TableCell>
                  <a
                    className={classes.link}
                    href={sourceLinkGenerator(item.assigned_by)}
                    target="_blank"
                    rel="noopener noreferrer">
                    {item.assigned_by}
                  </a>
                </TableCell>
              </TableRow>
            ))}
        </TableBody>
      </Table>
    </Paper>
  )
}

export default DisplayTable
