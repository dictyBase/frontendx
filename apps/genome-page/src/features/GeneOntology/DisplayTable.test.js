import React from "react"
import renderer from "react-test-renderer"
import DisplayTable from "./DisplayTable"

test("matching a snapshot of DisplayTable", () => {
  const data = [
    {
      term: "ATP-gated ion channel activity",
      evidence: "IMP",
      with: "",
      reference: "Sivaramakrishnan & Fountain (2012)",
      date: "21-02-2014",
      source: "DDB",
    },
    {
      term: "calcium ion transmembrane transport requires ATP",
      evidence: "IGI",
      with: "p2xA p2xC p2xD p2xE",
      reference: "Sivaramakrishnan & Fountain (2012)",
      date: "21-02-2014",
      source: "DDB",
    },
  ]
  const fetchData = () => {}

  const component = renderer.create(
    <DisplayTable data={data} fetchData={fetchData} />,
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
