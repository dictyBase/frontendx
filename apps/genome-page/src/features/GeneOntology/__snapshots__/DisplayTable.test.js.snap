// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matching a snapshot of DisplayTable 1`] = `
<div
  className="wrapper"
/>
`;
